File: src/II102.cpp
#include <iostream>
#include <numeric>
#include <unordered_map>
#include <vector>

using namespace std;

// 极其暴力使用hash的解法
class Solution {
public:
  int findTargetSumWays(vector<int>& nums, int target) {
    int n = nums.size(), m = accumulate(nums.begin(), nums.end(), 0);
    vector<unordered_map<int, int>> dp(n + 1);
    dp[0][0] = 1;
    for (int i = 0; i < n; ++i) {
      for (int j = -m; j <= m; ++j) {
        dp[i + 1][j] = 0;
        if (dp[i].count(j - nums[i])) {
          dp[i + 1][j] += dp[i][j - nums[i]];
        }
        if (dp[i].count(j + nums[i])) {
          dp[i + 1][j] += dp[i][j + nums[i]];
        }
      }
    }
    return dp[n].count(target) ? dp[n][target] : 0;
  }
};

File: src/II102.rs
struct Solution;

// sum-2*neg=target => neg=(sum-target)/2
impl Solution {
    pub fn find_target_sum_ways(nums: Vec<i32>, target: i32) -> i32 {
        let sm: i32 = nums.iter().sum();
        if sm < target || (sm - target) & 1 == 1 {
            return 0;
        }
        let neg = ((sm - target) / 2) as usize;
        let mut dp = vec![0; neg + 1];
        dp[0] = 1;
        for num in nums {
            let num = num as usize;
            for j in (num..=neg).rev() {
                dp[j] += dp[j - num];
            }
        }
        dp[neg]
    }
}
